File: kmz_points/workbook_facts.py
"""A normalised description of a workbook, so two writers can be compared.

The browser version writes its xlsx by hand rather than through openpyxl, so
the two files can never be byte-identical: they order zip entries differently,
express the same style in a different shape, and disagree about whitespace no
reader ever sees. Comparing bytes would fail on every run while proving
nothing, so what gets compared instead is what a person opening the file would
find: which sheets exist, what is in each cell, how it is typed and formatted,
what is merged, how wide the columns are and where the freeze sits.

Fonts, fills, colours and alignment are deliberately left out. That is a scope
decision, not an oversight. The two implementations model styling completely
differently -- openpyxl hangs a style object off each cell, the browser writer
registers a look once and stores an index -- so comparing them would compare
two encodings rather than two workbooks. The facts below are the ones that
decide what a cell actually shows.
"""

from __future__ import annotations

from typing import Any

from openpyxl import Workbook, load_workbook
from openpyxl.utils import get_column_letter
from openpyxl.utils.cell import coordinate_to_tuple


def _freeze_row(sheet: Any) -> int:
    """The last frozen row, or 0 where nothing is frozen.

    openpyxl states this as the top-left cell of the *unfrozen* region -- "A4"
    means rows 1 to 3 are frozen -- which is one off from the count anything
    comparing two sheets wants.
    """
    frozen = sheet.freeze_panes
    if not frozen:
        return 0
    return coordinate_to_tuple(str(frozen))[0] - 1


def _column_widths(sheet: Any) -> list[float]:
    """One width per column the sheet uses, in column order."""
    return [
        round(sheet.column_dimensions[get_column_letter(index)].width, 2)
        for index in range(1, sheet.max_column + 1)
    ]


def _cell_facts(cell: Any) -> dict[str, Any]:
    value = cell.value
    if isinstance(value, str):
        return {
            "row": cell.row,
            "col": cell.column,
            "value": value,
            "type": "s",
            "numberFormat": cell.number_format,
        }

    # Six decimals is far beyond anything the table carries and short of where
    # float64 starts showing its own representation noise, which is the only
    # difference two languages computing the same number produce. The zero
    # case flattens -0.0, which reads as 0 but is not equal to it under the
    # strict deep-equal on the other side of the comparison.
    number = round(float(value), 6)
    return {
        "row": cell.row,
        "col": cell.column,
        "value": 0.0 if number == 0 else number,
        "type": "n",
        "numberFormat": cell.number_format,
    }


def _sheet_facts(sheet: Any) -> dict[str, Any]:
    cells = [
        _cell_facts(cell)
        for row in sheet.iter_rows()
        for cell in row
        # An empty cell is written only to carry a fill, and fills are not
        # compared, so it says nothing about what the sheet contains.
        if cell.value is not None and cell.value != ""
    ]
    cells.sort(key=lambda cell: (cell["row"], cell["col"]))

    return {
        "name": sheet.title,
        "freezeRow": _freeze_row(sheet),
        "columnWidths": _column_widths(sheet),
        "merges": sorted(str(area) for area in sheet.merged_cells.ranges),
        "cells": cells,
    }


def workbook_facts(source: Any) -> dict[str, Any]:
    """Normalised description of a workbook, for comparing two writers.

    ``source`` is a path, an open binary stream, or an openpyxl Workbook.
    """
    book = source if isinstance(source, Workbook) else load_workbook(source)
    return {"sheets": [_sheet_facts(book[name]) for name in book.sheetnames]}

File: web/test/workbook-facts.mjs
/**
 * The browser side of the workbook comparison.
 *
 * Reduces the sheets buildBatchSheets produces to exactly the shape
 * kmz_points/workbook_facts.py reports for a workbook openpyxl wrote, so the
 * two can be compared without either side having to agree on bytes -- which
 * they never will, an openpyxl file and a hand-written one differing in zip
 * ordering, style encoding and whitespace no reader ever sees.
 *
 * Fonts, fills, colours and alignment are deliberately not included; see the
 * module docstring of workbook_facts.py for why.
 *
 * Lives in test/ rather than src/ because it is only ever used by the test
 * suite, and web/src/ is bundled into the single-file build.
 */

// Where xlsx.js starts its own number formats: ids below 164 are reserved by
// the format for built-ins, and 0 means the cell carries no format at all.
const FIRST_CUSTOM_FORMAT = 164;

const CELL_KEYS = ["row", "col", "value", "type", "numberFormat"];

function round(value, places) {
  const number = Number(value.toFixed(places));
  // -0 reads as 0 but is not equal to it under a strict deep-equal.
  return number === 0 ? 0 : number;
}

function numberFormatOf(styles, style) {
  const id = styles.cellFormats[style || 0].numberFormat;
  return id ? styles.numberFormats[id - FIRST_CUSTOM_FORMAT] : "General";
}

/** Value and type as xlsx.js would write them: finite numbers, else text. */
function typedValue(value) {
  if (typeof value === "number" && Number.isFinite(value)) {
    return { value: round(value, 6), type: "n" };
  }
  return { value, type: "s" };
}

function sheetFacts(sheet, styles) {
  const cells = [];
  sheet.rows.forEach((row, rowIndex) => {
    row.forEach((cell, columnIndex) => {
      // A cell with nothing in it is emitted only so a banner's fill covers
      // the row, and fills are not compared -- so, like the Python side, it
      // says nothing about what the sheet contains.
      if (!cell || cell.value === null || cell.value === undefined || cell.value === "") {
        return;
      }
      cells.push({
        row: rowIndex + 1,
        col: columnIndex + 1,
        ...typedValue(cell.value),
        numberFormat: numberFormatOf(styles, cell.style),
      });
    });
  });

  return {
    name: sheet.name,
    freezeRow: sheet.freezeRow,
    columnWidths: sheet.columnWidths.map((width) => round(width, 2)),
    merges: [...sheet.merges].sort(),
    cells,
  };
}

/** The facts of a batch's sheets, as workbook_facts.py would state them. */
export function workbookFacts({ sheets, look }) {
  return { sheets: sheets.map((sheet) => sheetFacts(sheet, look.styles)) };
}

function show(value) {
  return typeof value === "string" ? JSON.stringify(value) : String(value);
}

function describeCell(cell) {
  return cell ? `${show(cell.value)} (${cell.type}, ${cell.numberFormat})` : "nothing";
}

function listDifference(actual, expected, describe) {
  const shared = Math.min(actual.length, expected.length);
  for (let index = 0; index < shared; index += 1) {
    if (actual[index] !== expected[index]) {
      return describe(index, actual[index], expected[index]);
    }
  }
  if (actual.length === expected.length) return null;
  const index = shared;
  return describe(index, actual[index], expected[index]);
}

function cellsDifference(actual, expected) {
  const shared = Math.min(actual.length, expected.length);
  for (let index = 0; index < shared; index += 1) {
    const mine = actual[index];
    const theirs = expected[index];
    if (CELL_KEYS.every((key) => mine[key] === theirs[key])) continue;
    // Row and column are part of the comparison, so a cell one side wrote and
    // the other did not shows up here as a shift rather than as a wrong value.
    const at =
      mine.row === theirs.row && mine.col === theirs.col
        ? `row ${mine.row}, column ${mine.col}`
        : `row ${mine.row}, column ${mine.col} where Python has row ` +
          `${theirs.row}, column ${theirs.col}`;
    return `${at}: browser has ${describeCell(mine)}, Python has ${describeCell(theirs)}`;
  }

  if (actual.length === expected.length) return null;
  const extra = actual[shared] || expected[shared];
  return (
    `the browser wrote ${actual.length} cells and Python ${expected.length}; ` +
    `they first part ways at row ${extra.row}, column ${extra.col}: ` +
    `browser has ${describeCell(actual[shared])}, Python has ${describeCell(expected[shared])}`
  );
}

/**
 * Where two sets of facts first disagree, in one line, or null if they agree.
 *
 * A deep-equal on two workbooks of this size prints thousands of lines and
 * says nothing about which cell went wrong, which is the only thing anyone
 * reading the failure wants to know.
 */
export function firstDifference(actual, expected) {
  const names = (facts) => facts.sheets.map((sheet) => sheet.name).join(", ");
  if (actual.sheets.length !== expected.sheets.length) {
    return `the browser wrote sheets [${names(actual)}], Python [${names(expected)}]`;
  }

  for (const [index, mine] of actual.sheets.entries()) {
    const theirs = expected.sheets[index];
    const where = `sheet ${index + 1} (${theirs.name})`;

    if (mine.name !== theirs.name) {
      return `${where}: the browser calls it ${show(mine.name)}`;
    }
    if (mine.freezeRow !== theirs.freezeRow) {
      return `${where}: browser freezes ${mine.freezeRow} rows, Python ${theirs.freezeRow}`;
    }

    const widths = listDifference(
      mine.columnWidths,
      theirs.columnWidths,
      (at, ours, python) =>
        `${where}: column ${at + 1} is ${show(ours)} wide in the browser and ` +
        `${show(python)} in Python`,
    );
    if (widths) return widths;

    const merges = listDifference(
      mine.merges,
      theirs.merges,
      (at, ours, python) =>
        `${where}: merge ${at + 1} of ${theirs.merges.length} is ${show(ours)} in the ` +
        `browser and ${show(python)} in Python`,
    );
    if (merges) return merges;

    const cells = cellsDifference(mine.cells, theirs.cells);
    if (cells) return `${where}, ${cells}`;
  }

  return null;
}
